feat: Add support for Phi-4-mini

brb-nv opened the pull request:
This MR adds support for Phi-4-mini model.

brb-nv committed:
Add support for Phi-4-mini

Signed-off-by: Balaram Buddharaju <169953907+brb-nv@users.noreply.github.com>

brb-nv commented:
/bot run

kaiyux commented:
/bot run

niukuo commented:
[PR_Github #210](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/210/) [ run ] triggered by Bot

niukuo commented:
[PR_Github #210](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/210/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #222](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/222/) completed with status: 'SUCCESS'

amukkara commented on examples/phi/convert_checkpoint.py:
```diff
@@ -154,13 +154,13 @@ def args_to_quant_config(args: argparse.Namespace) -> QuantConfig:
         assert False, "Invalid model type"
 
-    is_phi3 = 'Phi3' in model_type or 'MoE' in model_type
+    is_phi3 = 'Phi3' in model_type or 'MoE' in model_type or 'Phi4MM' in model_type
```
nit unrelated to Phi4MM: change MoE filter to `PhiMoE in model_type`

brb-nv replied:
Done.

amukkara commented on tensorrt_llm/models/phi3/config.py:
```diff
@@ -91,9 +91,19 @@ def from_hugging_face(
             kwargs['dense_attention_every_n_layers'] = getattr(
                 hf_config, "dense_attention_every_n_layers", None)
             kwargs['norm_epsilon'] = hf_config.layer_norm_epsilon
+            kwargs['rotary_pct'] = getattr(hf_config, "partial_rotary_factor",
```
since `rotary_pct` and `tie_word_embeddings` are common for small and non-small variants, their initialization can be moved before or after the if else block

brb-nv replied:
Done.

amukkara commented on tensorrt_llm/models/phi3/convert.py:
```diff
@@ -114,14 +116,15 @@ def get_moe_weight(key, suffix):
             del weights[key]
 
-    if config.architecture == 'Phi3SmallForCausalLM':
+    if config.tie_word_embeddings or config.architecture == 'Phi3SmallForCausalLM':
```
is `config.tie_word_embeddings == True` for Phi3SmallForCausalLM? If so RHS of OR is redundant.

brb-nv replied:
It's actually False, Anurag.

symphonylyh commented on tensorrt_llm/tools/multimodal_builder.py:
```diff
@@ -1014,6 +1018,133 @@ def forward(self, pixel_values):
+        max_batch_size=args.max_batch_size,
+        engine_name=f"visual_encoder.engine",
+        dtype=torch.float16)
+    print("Built visual_encoder.engine.")
```
use logger

brb-nv replied:
Addressing this in https://github.com/NVIDIA/TensorRT-LLM/pull/3177.

symphonylyh commented on tensorrt_llm/tools/multimodal_builder.py:
```diff
@@ -1014,6 +1018,133 @@ def forward(self, pixel_values):
+        max_batch_size=args.max_batch_size,
+        engine_name=f"audio_encoder.engine",
+        dtype=torch.float16)
+    print("Built audio_encoder.engine.")
```
use logger

brb-nv replied:
Addressing this in https://github.com/NVIDIA/TensorRT-LLM/pull/3177.

symphonylyh commented on tests/integration/test_lists/test-db/l0_h100.yml:
are phi-4-mm tests missing in test_multimodal_general.py?

brb-nv replied:
Addressing this in https://github.com/NVIDIA/TensorRT-LLM/pull/3177.

amukkara commented on tensorrt_llm/runtime/multimodal_model_runner.py:
```diff
@@ -1371,6 +1533,40 @@ def get_visual_features(self, image, other_vision_inputs):
+
+        audio_embeds = audio_outputs[self.audio_output_names[0]]
+        return audio_embeds
+
```
can you create a general `get_features` routine and invoke it within `get_visual_features` and `get_audio_features` to avoid code repetition.

brb-nv replied:
Addressing this in https://github.com/NVIDIA/TensorRT-LLM/pull/3177.

brb-nv commented:
Hi @amukkara @symphonylyh, I'm descoping this MR to have only changes for Phi-4-mini model. Changes for Phi-4-MM will be done in this MR: https://github.com/NVIDIA/TensorRT-LLM/pull/3177

I've addressed all your comments (including those related to the multimodal model). However, current changes for Phi4MM need some overhauling (using Pytorch impl just like Phi3 Vision for encoders instead of TRT engine). I felt separating the two out is cleaner instead of blocking one for the other.

brb-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #804](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/804/) [ run ] triggered by Bot

brb-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #805](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/805/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #804](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/804/) [ run ] completed with state `ABORTED`

amukkara reviewed (approved):
reviewed changes after Phi4MM was removed. lgtm.

tensorrt-cicd commented:
[PR_Github #805](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/805/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #651](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/651/) completed with status: 'SUCCESS'

brb-nv commented:
/bot run

tensorrt-cicd commented:
[PR_Github #917](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/917/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #917](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/917/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #724](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/724/) completed with status: 'SUCCESS'

kaiyux commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #944](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/944/) [ reuse-pipeline ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #944](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/944/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #917](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/917/) for commit `179d06d`
